libvast/vast/system/datagram_source.hpp

```diff
@@ -0,0 +1,167 @@
+
+  using super = source_state<Reader>;
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  using super::super;
+
+  // -- member variables -------------------------------------------------------
+
+  /// Shuts down the stream manager when `true`.
+  bool done = false;
+};
+
+template <class Reader>
+using datagram_source_actor = caf::stateful_actor<datagram_source_state<Reader>,
+                                                  caf::io::broker>;
+
+/// An event producer.
+/// @tparam Reader The concrete source implementation.
+/// @param self The actor handle.
+/// @param reader The reader instance.
+template <class Reader>
+caf::behavior
+datagram_source(datagram_source_actor<Reader>* self,
+                uint16_t udp_listening_port, Reader reader,
+                typename source_state<Reader>::factory_type factory,
+                size_t table_slice_size) {
+  using namespace caf;
+  using namespace std::chrono;
+  // Try to open requested UDP port.
+  auto udp_res = self->add_udp_datagram_servant(udp_listening_port);
+  if (!udp_res) {
+    VAST_ERROR(self, "could not open port", udp_listening_port);
+    return {};
+  }
+  VAST_DEBUG(self, "starts listening at port", udp_res->second);
+  // Initialize state.
+  self->state.init(self, std::move(reader), std::move(factory));
+  // Spin up the stream manager for the source.
+  self->state.mgr = self->make_continuous_source(
+    // init
+    [=](caf::unit_t&) {
+      timestamp now = system_clock::now();
+      self->send(self->state.accountant, "source.start", now);
+    },
+    // get next element
+    [](caf::unit_t&, downstream<table_slice_handle>&, size_t) {
+      // nop, new slices are generated in the new_datagram_msg handler
+    },
+    // done?
+    [=](const caf::unit_t&) {
+      return self->state.done;
+    }
+  );
+  return {
+    [=](get_atom, schema_atom) -> result<schema> {
+      auto sch = self->state.reader.schema();
+      if (sch)
+        return *sch;
+      return sch.error();
+    },
+    [=](put_atom, const schema& sch) -> result<void> {
+      auto r = self->state.reader.schema(sch);
+      if (r)
+        return {};
+      return r.error();
+    },
+    [=](expression& expr) {
+      VAST_DEBUG(self, "sets filter expression to:", expr);
+      self->state.filter = std::move(expr);
+    },
+    [=](caf::io::new_datagram_msg& msg) {
+      // Check whether we can buffer more slices in the stream.
+      VAST_DEBUG(self, "got a new datagram of size", msg.buf.size());
+      auto& st = self->state;
+      auto capacity = st.mgr->out().capacity();
+      if (capacity == 0) {
+        VAST_WARNING(self, "has no capacity left in stream, dropping input!");
+        return;
+      }
+      // Extract events until the source has exhausted its input or until
+      // we have completed a batch.
+      auto start = steady_clock::now();
+      caf::arraybuf<> buf{msg.buf.data(), msg.buf.size()};
```

Comment:
Is `<>` still needed with C++17 template deduction?

Reply:
Template deduction doesn't help here, because the constructor arguments are always `(const char*, size_t)`.

Reply:
Ah right, makes sense.